cs/mineralpacket.go

```diff
@@ -279,33 +281,85 @@ func (packet *MineralPacket) estimateDamage(rules *Rules, player *Player, target
 }
 
 // Check if an uncaught PP packet will terraform the target planet's environment  (50% chance/100kT)
-func (packet *MineralPacket) checkTerraform(rules *Rules, player *Player, planet *Planet, uncaught int) {
+func (packet *MineralPacket) checkTerraform(rules *Rules, player *Player, planet *Planet, uncaught float64) {
 	if player.Race.Spec.PacketTerraformChance > 0 {
-		for uncaughtCheck := player.Race.Spec.PacketPermaTerraformSizeUnit; uncaughtCheck <= uncaught; uncaughtCheck += player.Race.Spec.PacketPermaTerraformSizeUnit {
-			if player.Race.Spec.PacketTerraformChance >= rules.random.Float64() {
-				terraformer := NewTerraformer()
+		terraformer := NewTerraformer()
+		t := terraform{}
+
+		// Evaluate each mineral type separately
+		for i, minType := range [3]CargoType{Ironium, Boranium, Germanium} {
+			mineral := int(math.Ceil(float64(packet.Cargo.GetAmount(minType)) * uncaught))
+			habType := HabType(i)
+			direction := 0
+			result := TerraformResult{}
+
+			// no mineral = no terraforming
+			if mineral == 0 {
+				continue
+			}
 
-				result := terraformer.TerraformOneStep(planet, player, nil, false)
-				if result.Terraformed() {
-					messager.planetPacketTerraform(player, planet, result.Type, result.Direction)
+			for uncaughtCheck := player.Race.Spec.PacketPermaTerraformSizeUnit; uncaughtCheck <= mineral; uncaughtCheck += player.Race.Spec.PacketPermaTerraformSizeUnit {
+				terraformChance := player.Race.Spec.PacketTerraformChance * math.Min(float64((mineral-uncaughtCheck)/player.Race.Spec.PacketPermaTerraformSizeUnit), 1)
```

Comment:
I'm not sure I understand this math. can you add a comment with an example?

Reply:
Actually good timing - I really ought to debug it further to see whether it actually works or not. (I'm still a bit sketchy on the finer details of loops in golang in terms of what executes when, etc etc)
What it's SUPPOSED to do is multiply the terraform chance by the amount of mineral actually present (so if there's 350kT minerals in total, the 4th check has its probability multiplied by 0.5 due to having 50/100kT of minerals)
If you think of a better implementation feel free to suggest it.